src/polyfills/custom-elements.js

```diff
@@ -0,0 +1,599 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @typedef {{
+ *   promise: !Promise<undefined>,
+ *   resolve: function(),
+ * }}
+ */
+let DeferredDef;
```

Comment:
None yet. I've manually tested, and was going to build this into a fully fledged node module with tests there.

Note that the our regular tests are going to run with this polyfill instead of `document.registerElement`. If that's not enough, I can begin the test suite in this repo and extract all of it out later?

Reply:
Have you actually tried running the tests with the experiment on?

Reply:
- https://github.com/ampproject/amphtml/pull/17205/files#diff-8f8103e0f4e0052eab727bfd5c21e090R107
- https://github.com/ampproject/amphtml/pull/17205/files#diff-1179de12a95783d46f8bf0056a156c74R28

I can change the experiment check to test for `getMode().test` too, if we want to ensure all tests run with it.

Reply:
I think this is missing most integration tests that use a fully built AMP.

Reply:
Made it so it always runs in tests.

Reply:
But not yet in integration tests, right?

Reply:
No, all test run with the new polyfill. At the [moment](https://github.com/ampproject/amphtml/blob/master/build-system/tasks/runtime-test.js#L76-L86), this includes:

- `SL_Chrome_67`
- `SL_Firefox_61`
- `SL_Safari_11`

We can enable more saucelabs browsers, but I don't think that's absolutely necessary until we're ready to fully launch the experiment?